lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,625 @@
+      context, "sifive.enterprise.grandcentral.ReferenceDataTapKey");
+  StringAttr internalKeyClass = StringAttr::get(
+      context, "sifive.enterprise.grandcentral.DataTapModuleSignalKey");
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+class GrandCentralTapsPass : public GrandCentralTapsBase<GrandCentralTapsPass> {
+  void runOnOperation() override;
+};
+
+void GrandCentralTapsPass::runOnOperation() {
+  auto circuitOp = getOperation();
+  LLVM_DEBUG(llvm::dbgs() << "Running the GCT Data Taps pass\n");
+
+  // Here's a rough idea of what the Scala code is doing:
+  // - Gather the `source` of all `keys` of all `DataTapsAnnotation`s throughout
+  //   the design.
+  // - Convert the sources, which are specified on modules, to the absolute
+  //   paths in all instances. E.g. module M tap x will produce a.x and b.x if
+  //   there are two instances a and b of module M in the design.
+  // - All data tap keys are specified on black box ports.
+  // - The code then processes every DataTapsAnnotation separately as follows
+  //   (with the targeted blackbox and keys):
+  // - Check for collisions between SV keywords and key port names (skip this).
+  // - Find all instances of the blackbox, but then just pick the first one (we
+  //   should probably do this for each?)
+  // - Process each key independently as follows:
+  // - Look up the absolute path of the source in the map. Ensure it exists and
+  //   is unambiguous. Make it relative to the blackbox instance path.
+  // - Look up the port on the black box.
+  // - Create a hierarchical SV name and store this as an assignment to the
+  //   blackbox port.
+  //   - DeletedDataTapKey: skip and don't create a wiring
+  //   - LiteralDataTapKey: just the literal
+  //   - ReferenceDataTapKey: relative path with "." + source name
+  //   - DataTapModuleSignalKey: relative path with "." + internal path
+  // - Generate a body for the blackbox module with the signal mapping
+
+  // Gather some string attributes in the context to simplify working with the
+  // annotations.
+  Strings strings(&getContext());
+
+  // Gather a list of extmodules that have data or mem tap annotations to be
+  // expanded.
+  SmallVector<AnnotatedExtModule, 4> modules;
+  for (auto &op : *circuitOp.getBody()) {
+    auto extModule = dyn_cast<FExtModuleOp>(&op);
+    if (!extModule)
+      continue;
+
+    // Go through the module ports and collect the annotated ones.
+    AnnotatedExtModule result{extModule, {}, {}};
+    for (unsigned argNum = 0, e = extModule.getNumArguments(); argNum < e;
+         ++argNum) {
+      // Go through all annotations on this port and add the data tap key and
+      // mem tap ones to the list.
+      auto annos = AnnotationSet(
+          extModule.getArgAttrOfType<ArrayAttr>(argNum, strings.fannos));
```

Comment:
Please add a helper method on AnnotationSet to pull the magic "firrtl.annotations" string into FIRRTLAnnotations.cpp, and s/arg/port/.  Maybe something like:

`auto annos = AnnotationSet::getPortAttrs(extModule, argNum);`

?

Reply:
This has been added in the meantime; switched over to that.